gpflow/likelihoods.py

```diff
@@ -583,3 +583,136 @@ def conditional_variance(self, F):
         E_y = tf.matmul(phi, Ys)
         E_y2 = tf.matmul(phi, tf.square(Ys))
         return tf.reshape(E_y2 - tf.square(E_y), tf.shape(F))
+
+
+class MonteCarloLikelihood(Likelihood):
+    def __init__(self, name=None):
+        super().__init__(name)
+        self.num_monte_carlo_points = 100
+        del self.num_gauss_hermite_points
+
+    def _mc_evals(self, func, Fmu, Fvar, Y=None, epsilon=None):
+        N, D = tf.shape(Fmu)[0], tf.shape(Fvar)[1]
+        if epsilon is None:
+            epsilon = tf.random_normal((self.num_monte_carlo_points, N, D), dtype=settings.float_type)
+        mc_x = Fmu[None, :, :] + tf.sqrt(Fvar[None, :, :]) * epsilon
+        mc_Xr = tf.reshape(mc_x, (self.num_monte_carlo_points * N, D))
+
+        if Y is not None:
+            mc_Yr = tf.tile(Y, [self.num_monte_carlo_points, 1])  # broadcast Y to match X
+            f_eval = tf.reshape(func(mc_Xr, mc_Yr),
+                                (self.num_monte_carlo_points, N, 1))  # S x N x 1
+
+        else:
+            f_eval = tf.reshape(func(mc_Xr),
+                                (self.num_monte_carlo_points, N, D))
+
+        return f_eval
+
+    def predict_mean_and_var(self, Fmu, Fvar, epsilon=None):
+        r"""
+        Given a Normal distribution for the latent function,
+        return the mean of Y
+
+        if
+            q(f) = N(Fmu, Fvar)
+
+        and this object represents
+
+            p(y|f)
+
+        then this method computes the predictive mean
+
+           \int\int y p(y|f)q(f) df dy
+
+        and the predictive variance
+
+           \int\int y^2 p(y|f)q(f) df dy  - [ \int\int y^2 p(y|f)q(f) df dy ]^2
+
+        Here, we implement a default Monte Carlo routine.
+        """
+        mc_cm = self._mc_evals(self.conditional_mean, Fmu, Fvar, epsilon=epsilon)
+
+        mean, var = tf.nn.moments(mc_cm, [0])
+        unbiased_var = var * self.num_monte_carlo_points / (self.num_monte_carlo_points - 1)
```

Comment:
What's the reasoning behind this? Could you add a small comment? All it needs is something that explains what `tf.nn.moments()` returns, and then why the factor makes it unbiased.

Reply:
done